### PR TITLE
[FIXED JENKINS-45436] Add an API to assist deriving human names from URLs

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-45436](https://issues.jenkins-ci.org/browse/JENKINS-45436)

@reviewbybees

### CODE REVIEW

File: src/main/java/jenkins/scm/api/SCMName.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+ *
+ * @since 2.2.0
+ */
+public final class SCMName {
+
+    /**
+     * Utility class
+     */
+    private SCMName() {
+        throw new IllegalAccessError("Utility class");
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/scm/api/SCMName.java]: g1333397188557621642

File: src/main/java/jenkins/scm/api/SCMName.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    /**
+     * Makes best effort to guess a "sensible" display name from the hostname in the URL.
+     *
+     * @param url the URL.
+     * @return the display name or {@code null}
+     */
+    public static String fromUrl(@NonNull String url, @CheckForNull String... ignoredPrefixes) {
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/scm/api/SCMName.java]: Annotate return value?

File: src/main/java/jenkins/scm/api/SCMName.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+     * Makes best effort to guess a "sensible" display name from the hostname in the URL.
+     *
+     * @param url the URL.
+     * @return the display name or {@code null}
+     */
+    public static String fromUrl(@NonNull String url, @CheckForNull List<String> ignoredPrefixes) {
+        try {
+            return innerFromUrl(url, ignoredPrefixes);
+        } catch (LinkageError e) {
+            return null;
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/scm/api/SCMName.java]: same

File: src/main/java/jenkins/scm/api/SCMName.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                            hostName = hostName.substring(prefix.length() + 1);
+                            break;
+                        }
+                    }
+                }
+            }
+            if (StringUtils.isNotBlank(hostName)) {
+                hostName = IDN.toUnicode(hostName).replace('.', ' ').replace('-', ' ');
+            }
+        } catch (MalformedURLException e) {
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/scm/api/SCMName.java]: I would suggest reducing the scope of try/catch to `new URL(url)`

Comment (oleg-nenashev) [inline src/main/java/jenkins/scm/api/SCMName.java]: Or maybe it makes sense to catch `Exception` and to log it on the CONFIG level

File: src/main/java/jenkins/scm/api/SCMName.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+import java.util.List;
+import java.util.Locale;
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * Utility class to help with naming SCM related things.
+ *
+ * @since 2.2.0
+ */
+public final class SCMName {
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/scm/api/SCMName.java]: The name is fine for me, at some point we may also need an `SCMNameResolver` extension point

File: src/main/java/jenkins/scm/api/SCMName.java (hunk 1 above)
Comment (oleg-nenashev) [inline src/main/java/jenkins/scm/api/SCMName.java]: It is reasonable to be protected from Groovy magic though

File: src/main/java/jenkins/scm/api/SCMName.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+    /**
+     * Makes best effort to guess a "sensible" display name from the hostname in the URL.
+     *
+     * @param url the URL.
+     * @return the display name or {@code null}
+     * @throws LinkageError if Guava changes their API that we have depended on.
+     */
+    @CheckForNull
+    private static String innerFromUrl(@NonNull String url, List<String> ignoredPrefixes) throws LinkageError {
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/scm/api/SCMName.java]: CheckForNull for `ignoredPrefixes ` would not hurt

File: src/main/java/jenkins/scm/api/SCMName.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    /**
+     * Utility class
+     */
+    private SCMName() {
+        throw new IllegalAccessError("Utility class");
+    }
+
+    /**
+     * Makes best effort to guess a "sensible" display name from the hostname in the URL.
+     *
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/scm/api/SCMName.java]: Would be nice to provide examples in Javadoc

File: src/test/java/jenkins/scm/api/SCMNameTest.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    @Test
+    public void given__url_with_punycode__when__naming__then__hostname_is_decoded() throws Exception {
+        assertThat(SCMName.fromUrl("http://xn--e1afmkfd.xn--p1ai/"), is("пример"));
+    }
+
+    @Test
+    public void given__url_with_idn__when__naming__then__punycode_is_roundtripped() throws Exception {
+        assertThat(SCMName.fromUrl("http://\u043F\u0440\u0438\u043C\u0435\u0440.\u0440\u0444" /*пример.рф*/),
```
Comment (stephenc) [inline src/test/java/jenkins/scm/api/SCMNameTest.java]: I hope I have translated correctly @oleg-nenashev ;-)

**Comment by rsandell** [comment]
:bee:

**Pull request merged**